text/3389-manifest-lint.md

```diff
@@ -0,0 +1,250 @@
+all workspace members.  However, `[patch]` and friends are because they affect
+the resolver / `Cargo.toml` and so they can only operate at the workspace
+level.  `[lints]` is more like `[dependencies]` in being something that applies
+at the package level but we want shared across workspaces.
+
+Instead of traditional workspace inheritance where there is a single value to
+inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
+which defines presets and the user could do `lints.<preset> = true`.  The user
+could then name them as they wish to avoid collision with rustc lints.
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
+
+Should we only apply/fingerprint lints for the appropriate tool?  For example,
+we would not include and fingerprint `clippy::` lints when running builds,
+allowing them to change without forcing a rebuild.  We likely already need to
+be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
+this isn't much more of a step.
+
+How do we allow controling precedence between lints and lint groups?  We are
+using a TOML table with the keys as lint names which does not allow controlling
+ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
+precedence between levels that the user can't control.
+
+# Future possibilities
```

Comment:
One future possibility I'd like to see here would be an option to link to a file that defines lint levels. A lot of people would prefer an extra file (me included) that can just be copied across projects. Instead of always copying the list out of the `Cargo.toml`, people would then just have to mention the file in the `Cargo.toml` and can copy it around (or submodule it from a central repo).

This would be a good compromise of putting the configuration inside `Cargo.toml` instead of a `lints.toml` file.

I would keep this as a future possibility, because I don't consider this important for the MVP.

Reply:
Yeah, that would be nice.

Reply:
I overlooked that because I already have to diff `Cargo.toml` files between all my projects as I percolate out changes in my practices.

I've now added it to future possibilities, pointing to `package.license` as precedence.